# Classifying immune T-cell types

## Introduction
Transcriptome data typically result in large datasets containing tens of thousands of genes. The goal of analyzing transcriptome datasets are too find
a subset of genes with expression modulations that contribute to physiological change in the model organism. One such approach is gene differential
expression analysis, which uses linear regression models on each gene to identify changes in expression between biological conditions. These models
are created with the gene read counts as the dependent variable and biological condition as the independent variable. However, this approach is
overreliant on the p-value to determine if a gene is differentially expressed and suffers when biological variability is high resulting in few genes
being identified. In addition, this is computationally intensive because a model must be fit for every individual gene.

I suggest an alternative approach to identifying a subset of biological relevant genes that utilizes machine learning methods. This approach instead
fits a single machine learning model using the biological condition as the dependent variable (aka, target variable in machine learning speak) and
takes each gene and its read counts as its own independent variable (aka, feature variable). The goal of this approach is to use the transcriptome
data to make predictions/classifications on the biological condition. This uses machine learning models such as LASSO regression and ensemble methods
that inherently filter out feature variables not contributing to the prediction/classification. The resulting models will contain features/genes
that can be extracted resulting in a subset of genes significantly contributing to a biological condition.

## Dataset
To demonstrate this approach, I will be using a dataset obtained from the National Library of Medicine (NCBI) Gene Expression Omnibus (GEO) database
([Accession number: GSE107011](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107011)) under an
[open-license](https://www.ncbi.nlm.nih.gov/geo/info/disclaimer.html) with no copyright restrictions. The data was generated by
[Monaco et al., 2019](https://www.cell.com/cellreports/fulltext/S2211-1247(19)30059-2?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2211124719300592%3Fshowall%3Dtrue#secsectitle0135)
for their paper entitled “RNA-Seq Signatures Normalized by mRNA Abundance Allow Absolute Deconvolution of Human Immune Cell Types”.

The transcriptome data contains TPM (Transcripts Per Million) normalized read counts for 127 total samples from 8 different immune cell types in 13
healthy Singaporean individuals. Cell types include B/T cells, Dendiritic Cells (DC), Natural Killer (NK) cells, granulocytes, monocytes, progenitor
cells, and Peripheral Blood Mononuclear Cells (PBMCs). To simplify predictions/classifications, I have grouped these samples into two immune types:
T-cells (58 samples) and non T-cells (69 samples).

## Pipeline of analysis
There are two main portions to this pipeline:

1. Gene module creation with WGCNA
2. Machine learning classifiers

### 1. WGCNA
This portion of the pipeline attempts to create an initial subset of genes that are associated with the trait or biological condition of interest
(in this case T-cells). It does this through creating modules/clusters of genes using
[WGCNA](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559) and correlates them to the trait of interest. Hub genes from
these identified gene modules are then extracted as features to be used in the machine learning models.

### 2. Machine learning
Various machine learning models including LASSO logistic regression, random forest classifier, and adaptive boost classifier are trained and tested
to obtain the final subset of genes that are significantly contributing to the prediction/classification of T-cells. This subset of genes can be further
analyzed for their biological functions using various enrichment analyses.

The figure below provides an overall summary of how the scripts were used along with their data inputs and outputs.

![alt text](https://github.com/dswede43/bioinformatics_analyses/blob/b5b87813d602f4981f243eb30b61d8a0b1f20f32/Machine%20learning/Immune%20cell%20classification/analysis_pipeline.JPG)

## Results
The initial transcriptome dataset contained read counts 58,174 genes. After filtering out genes with low read counts (ie. genes with an average TPM
less than 10), this resulted in 10,247 genes to be used in WGCNA.

### 1. WGCNA
WGCNA clustered the 10,247 genes into 16 different modules as shown by the dendrogram below.

![alt text](https://github.com/dswede43/bioinformatics_analyses/blob/2fd8490619a38ab629bad228ce3035503553c4f3/Machine%20learning/Immune%20cell%20classification/Visualizations/module_dendrogram.pdf)

Of these 16 gene modules, 10 of them significantly correlated (FDR adjusted p-value < 0.001) with the trait of interest (T-cells).

![alt text](https://github.com/dswede43/bioinformatics_analyses/blob/2fd8490619a38ab629bad228ce3035503553c4f3/Machine%20learning/Immune%20cell%20classification/Visualizations/ME_trait_cor_heatmap.pdf)

From these 10 modules, 826 genes were identified as hub genes and used as the feature variables for the machine learning models.

### 2. Machine learning


## Conclusion
